### PR TITLE
Named arguments syntax with function definitions

### LINKED ISSUE #2365: Function definitions in named arguments
Allow e.g.:
```
mkRec (
  f : A -> A  
    | c := d
    | d := e;
)
```

### PR DESCRIPTION
**Pull request by lukaszcz**
* Closes #2365
* Implements the syntax `f@{x1 := def1; ...; xn := defn}` and `f@?{x1 := def1; ..; xn := defn}`. Each definition inside the `@{..}` is an ordinary function definition. The `@?` version allows partial application (not all explicit named arguments need to be provided). This subsumes the old record creation syntax.

### CODE REVIEW

File: src/Juvix/Compiler/Internal/Translation/FromConcrete.hs [hunk 1]
```diff
@@ -764,6 +765,74 @@ goExpression = \case
...
+                      }
+                  goArg :: Symbol -> NamedArgument 'Scoped
+                  goArg sym =
+                    NamedArgument
+                      { _namedArgName = sym,
+                        _namedArgAssignKw = Irrelevant dummyKw,
+                        _namedArgValue = Concrete.ExpressionIdentifier $ ScopedIden name Nothing
+                      }
+                    where
+                      name = over S.nameConcrete NameUnqualified $ fromJust $ find ((== sym) . (^. S.nameConcrete)) args
```
Comment (paulcadman) [inline src/Juvix/Compiler/Internal/Translation/FromConcrete.hs:832]: As `sym` ultimately comes from `args` itself, could `sym` and `name` be associated somehow during processing so it doesn't need to be obtained using `fromJust . find` at the end?

File: src/Juvix/Compiler/Internal/Translation/FromConcrete.hs [hunk 2]
```diff
@@ -764,6 +765,74 @@ goExpression = \case
...
+          createArgumentBlocks :: [NameBlock 'Scoped] -> [ArgumentBlock 'Scoped]
+          createArgumentBlocks sblocks = snd $ foldr goBlock (args0, []) sblocks
+            where
+              args0 :: HashSet S.Symbol = HashSet.fromList $ fmap (^. namedArgumentNewFunDef . signName) (toList appargs)
+              goBlock :: NameBlock 'Scoped -> (HashSet S.Symbol, [ArgumentBlock 'Scoped]) -> (HashSet S.Symbol, [ArgumentBlock 'Scoped])
+              goBlock NameBlock {..} (args, blocks)
+                | null names = (args', blocks)
+                | otherwise = (args', block' : blocks)
+                where
+                  names =
```
Comment (paulcadman) [inline src/Juvix/Compiler/Internal/Translation/FromConcrete.hs:812]: `names` -> `namesInBlock`?

Comment (lukaszcz) [reply src/Juvix/Compiler/Internal/Translation/FromConcrete.hs:812]: done

File: src/Juvix/Compiler/Internal/Translation/FromConcrete.hs [hunk 3]
```diff
@@ -764,6 +765,74 @@ goExpression = \case
...
+                    HashSet.intersection
+                      (HashSet.fromList $ HashMap.keys _nameBlock)
+                      (HashSet.map (^. S.nameConcrete) args)
+                  args' = HashSet.filter (not . flip HashSet.member names . (^. S.nameConcrete)) args
+                  nargs = nonEmpty' $ map goArg (toList names)
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Internal/Translation/FromConcrete.hs:817]: done

File: src/Juvix/Compiler/Internal/Translation/FromConcrete.hs (hunk 1 above)
Comment (lukaszcz) [reply src/Juvix/Compiler/Internal/Translation/FromConcrete.hs]: I created a HashMap associating symbols with names

**Review by paulcadman — APPROVED** [review]

**Comment by jonaprieto** [comment]
This must be documented, please. I just saw this.

**Pull request merged**